File: ts/src/money/money_test.ts
import assert from "assert";

class Money {
  amount: number;
  currency: string;
  divide(divider: number) {
    return new Money(this.amount / divider, this.currency);
  }
  times(multiplier: number): Money {
    return new Money(this.amount * multiplier, this.currency);
  }
  constructor(amount: number, currency: string) {
    this.amount = amount;
    this.currency = currency;
  }
}

/**
 *
 * Chapter 1: The money problem
 *
 * */

const fiveDollars = new Money(5, "USD");
const tenDollars = new Money(10, "USD");

assert.deepStrictEqual(fiveDollars.times(2), tenDollars);

/**
 *
 * Chapter 2: Multi currency money
 *
 * */

const fiveEuros = new Money(5, "EUR");
const tenEuros = new Money(10, "EUR");

assert.deepStrictEqual(fiveEuros.times(2), tenEuros);

/**
 *
 * Chapter 2: Division
 *
 * */

const originalMoney = new Money(4002, "KRW");
const actualMoneyAfterDivision = originalMoney.divide(4);
const expectedMoneyAfterDivision = new Money(1000.5, "KRW");

assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision);

File: go/money/src/money_test.go
package main

import "testing"

type Money struct {
	amount   float64
	currency string
}

func (d Money) Times(multiplier int) Money {
	return Money{
		amount:   d.amount * float64(multiplier),
		currency: d.currency,
	}
}

func (d Money) Divide(divisor int) Money {
	return Money{
		amount:   d.amount / float64(divisor),
		currency: d.currency,
	}
}

func assertEquals(t *testing.T, expected Money, actual Money) {
	if expected != actual {
		t.Errorf("Expected %+v Got %+v", expected, actual)
	}
}

func TestMultiplication(t *testing.T) {
	fiver := Money{
		amount:   5,
		currency: "USD",
	}

	actualMoneyBeforeMultiplication := fiver.Times(2)
	expectedMoneyAfterMultiplication := Money{
		amount:   10,
		currency: "USD",
	}
	assertEquals(t, actualMoneyBeforeMultiplication, expectedMoneyAfterMultiplication)
}

func TestMultiplicationInEuros(t *testing.T) {
	tenEuros := Money{
		amount:   10,
		currency: "EUR",
	}

	actualMoneyAfterMultiplication := tenEuros.Times(2)
	expectedMoneyAfterMultiplication := Money{
		amount:   20,
		currency: "EUR",
	}

	assertEquals(t, actualMoneyAfterMultiplication, expectedMoneyAfterMultiplication)
}

func TestDivision(t *testing.T) {
	originalMoney := Money{amount: 4002, currency: "KRW"}
	actualMoneyAfterDivision := originalMoney.Divide(4)
	expectedMoneyAfterDivision := Money{amount: 1000.5, currency: "KRW"}

	assertEquals(t, actualMoneyAfterDivision, expectedMoneyAfterDivision)
}
